service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/controller/ReportController.java

```diff
@@ -29,13 +38,32 @@
 
     @PostMapping(value = "/report", consumes = { TEXT_XML_VALUE, APPLICATION_XML_VALUE })
     @ResponseStatus(value = ACCEPTED)
-    public void postReport(@RequestBody String reportXml) throws JsonProcessingException {
+    public void postReport(@RequestBody String reportXml) {
         try {
-            POCDMT000002UK01ClinicalDocument1 clinicalDocument = extractClinicalDocument(reportXml);
+            SAXReader reader = new SAXReader();
+            Document document = reader.read(new StringReader(reportXml));
+
+            String distributionEnvelope = getDistributionEnvelope(document);
+            //TODO: this message id will be used in response
+            String messageId = getMessageId(document);
+
+            POCDMT000002UK01ClinicalDocument1 clinicalDocument = extractClinicalDocument(distributionEnvelope);
             validate(clinicalDocument);
+
             encounterReportService.transformAndPopulateToGP(clinicalDocument);
-        } catch (XmlException e) {
+        } catch (XmlException | DocumentException e) {
             throw new ResponseStatusException(BAD_REQUEST, e.getMessage());
         }
     }
+
+    private String getMessageId(Document document) {
+        Node headerNode = document.selectSingleNode("//*[local-name()='Header']");
+        Node messageIdNode = headerNode.selectSingleNode("//*[local-name()='MessageID']");
```

Comment:
Isn't it enough to search for MessageID element in RootElement?

Reply:
it is :)